ui/src/app/components/invalid-billing-banner.spec.tsx

```diff
@@ -0,0 +1,198 @@
+
+describe('InvalidBillingBanner', () => {
+  const load = jest.fn();
+  const reload = jest.fn();
+  const updateCache = jest.fn();
+  const warningThresholdDays = 5; // arbitrary
+  const me = ProfileStubVariables.PROFILE_STUB.username;
+  const someOneElse = 'someOneElse@fake-research-aou.org';
+
+  const component = (
+    signatureState: DuccSignatureState = DuccSignatureState.UNSIGNED
+  ) =>
+    render(
+      <MemoryRouter>
+        <InvalidBillingBanner
+          {...{ signatureState }}
+          hideSpinner={() => {}}
+          showSpinner={() => {}}
+        />
+      </MemoryRouter>
+    );
+
+  beforeEach(() => {
+    registerApiClient(ProfileApi, new ProfileApiStub());
+
+    // Do I need this?
+    reload.mockImplementation(async () => {
+      const newProfile = await profileApi().getMe();
+      profileStore.set({ profile: newProfile, load, reload, updateCache });
+    });
+
+    profileStore.set({
+      profile: ProfileStubVariables.PROFILE_STUB,
+      load,
+      reload,
+      updateCache,
+    });
+    serverConfigStore.set({
+      config: {
+        ...defaultServerConfig,
+        initialCreditsExpirationWarningDays: warningThresholdDays,
+      },
+    });
+  });
+
+  const setupWorkspace = (
+    exhausted: boolean,
+    expired: boolean,
+    expiringSoon: boolean,
+    ownedByMe: boolean
+  ) => {
+    // Set expiration date to be in the past if expired, in the future if not.
+    // If expiring soon, set it to be within the warning threshold, and just outside otherwise.
+    // Expired and expiringSoon are mutually exclusive.
+    const daysUntilExpiration = expired
+      ? -1
+      : warningThresholdDays + (expiringSoon ? -1 : 1);
+    currentWorkspaceStore.next({
+      ...workspaceDataStub,
+      initialCredits: {
+        exhausted,
+        expired,
+        expirationEpochMillis: plusDays(Date.now(), daysUntilExpiration),
+      },
+      creator: ownedByMe ? me : someOneElse,
+    });
+  };
+
+  /* All banners have "initial credits" in the text. Banner text can have one or more links in it.
+   * React Testing Library has a hard time finding text that is split across multiple elements
+   * (like text and links), so we can't use getByText to find the banner text. Instead, this
+   * function will return the textContent of the element that contains the banner text. This will
+   * include the plain text and the text found in the links.
+   */
+
+  const getBannerText = () =>
+    screen.getAllByText(/initial credits/).pop().textContent;
+
+  const setProfileExtensionEligibility = (isEligible: boolean) => {
+    profileStore.set({
+      profile: {
+        ...ProfileStubVariables.PROFILE_STUB,
+        eligibleForInitialCreditsExtension: isEligible,
+      },
+      load,
+      reload,
+      updateCache,
+    });
+  };
+
+  it('should show expiring soon banner to user who created the workspace', async () => {
+    setupWorkspace(false, false, true, true);
```

Comment:
I'd prefer to see these made explicit, like
```
    const exhausted = false;
    const expired = false;
    const expiringSoon = true;
    const ownedByMe = true;
    setupWorkspace(exhausted, expired, expiringSoon, ownedByMe);
```

Reply:
Great idea